### PR TITLE
Editpage

### PR DESCRIPTION
**Pull request by FabianAlleblas**
I added a edit function, so a user can edit his entries.
I'm not happy with the update function in the backend, but it works and I don't know how to write it better 🤷‍♂️ ...

The Edit-Form gets the current data as initial value.

### CODE REVIEW

File: backend/src/Controller/PlantController.php [hunk 1]
```diff
@@ -38,8 +35,32 @@ public function createPlant(Request $request, PlantRepository $plantRepository,
...
+        $validatedData = json_decode($serializer->serialize($newData, 'json'), true);
+        
+        $plantRepository->updatePlant($plant, $validatedData);
+        return $this->jsonResponse($plant, $serializer);
+     }
+
+    private function jsonResponse($data, $serializer): JsonResponse {
         return new JsonResponse(
-            $serializer->serialize($plant, 'json'),
+            $serializer->serialize($data, 'json', [AbstractObjectNormalizer::SKIP_NULL_VALUES => true]),
```
Comment (janabremer) [inline backend/src/Controller/PlantController.php:63]: nice 👌

File: backend/src/Repository/PlantRepository.php [hunk 2]
```diff
@@ -25,6 +25,18 @@ public function savePlant(Plant $plant): Plant {
         return $plant;
     }
 
+    public function updatePlant(Plant $plant, $newData): Plant{
```
Comment (janabremer) [inline backend/src/Repository/PlantRepository.php:28]: looks good to me 🤷‍♀️

File: frontend/src/App.js [hunk 3]
```diff
@@ -1,11 +1,16 @@
...
 import DetailPage from './Pages/DetailPage/DetailPage'
 import PlantListPage from './Pages/Plantlistpage/PlantListPage'
 
 function App() {
-  const { plantList, savePlantData } = usePlantList()
+  const { plantList, savePlantData, updatePlantData } = usePlantList()
+
+  if (!plantList) {
+    return <div>loading...</div>
+  }
```
Comment (janabremer) [inline frontend/src/App.js:13]: good idea

File: frontend/src/services/getData.js [hunk 4]
```diff
@@ -3,8 +3,8 @@ const baseUrl = 'http://urbanplants.local/plant'
 export default async function getData() {
```
Comment (janabremer) [inline frontend/src/services/getData.js:3]: "getPlants" might be more semantic

File: frontend/src/services/getData.js [hunk 5]
```diff
@@ -3,8 +3,8 @@ const baseUrl = 'http://urbanplants.local/plant'
...
   try {
     const response = await fetch(baseUrl)
-    const data = response.json()
-    return data
+    const data = await response.json()
+    return await data
   } catch (error) {
     return { error: 'The server is down! :(' }
```
Comment (janabremer) [inline frontend/src/services/getData.js:9]: oh :(

File: frontend/src/services/updateData.js [hunk 6]
```diff
@@ -0,0 +1,22 @@
+const baseUrl = 'http://urbanplants.local/plant'
+
+export default async function postData(data, id) {
```
Comment (janabremer) [inline frontend/src/services/updateData.js:3]: same here: postPlant or updatePlant
And is there a reason, you use different names for file and function?

**Review by janabremer — APPROVED** [review]

**Review by janabremer — COMMENTED** [review]
nice work, looks good to me! :)

File: frontend/src/services/getData.js (hunk 5 above)
Comment (FabianAlleblas) [reply frontend/src/services/getData.js:9]: :(

File: backend/src/Controller/PlantController.php [hunk 7]
```diff
@@ -38,8 +35,32 @@ public function createPlant(Request $request, PlantRepository $plantRepository,
...
+        $newData = $serializer->deserialize($request->getContent(), Plant::class, 'json');
+        $validationResult = $validator->validate($newData);
+        if ($validationResult->count() !== 0) {
+            return new JsonResponse(
+                ["error" => "Plant data invalid."],
+                JsonResponse::HTTP_BAD_REQUEST
+            );
+        }
+
+        $validatedData = json_decode($serializer->serialize($newData, 'json'), true);
```
Comment (web-devil) [inline backend/src/Controller/PlantController.php:55]: wieso?

Comment (web-devil) [reply backend/src/Controller/PlantController.php:55]: das sollte doch eine Plant sein und kein stdClass

File: frontend/src/Components/Forms/EditPlantForm.js [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+        <FormButton>Update</FormButton>
+        <FormButton onClick={handleCancel} secondaryStyle>
+          Cancel
+        </FormButton>
+        <p>* required</p>
+      </ButtonWrapper>
+    </Form>
+  )
+
+  function handleSubmit(e) {
```
Comment (web-devil) [inline frontend/src/Components/Forms/EditPlantForm.js:60]: event anstelle von e
(meaningful variable name)

**Review by web-devil — CHANGES_REQUESTED** [review]

File: backend/src/Controller/PlantController.php (hunk 7 above)
Comment (FabianAlleblas) [reply backend/src/Controller/PlantController.php:55]: plant in currentPlant und newData in newPlant, alright :)

**Pull request merged**